DOC: How to verify your first NumPy bug (#17851)
numpy/numpy: doc/source/user/how-to-verify-bug.rst

Create a new directory, enter into it and set up a virtual environment using
your preferred method. For example, this is how to do it using `virtualenv`:

::

    virtualenv venv_np_bug
    source venv_np_bug/bin/activate

******************************************************************************
2. Install the NumPy version mentioned in the bug report
******************************************************************************

If the bug was more complex, you may need to build the NumPy version they
referenced :ref:`from source <building-from-source>`. But for this bug,
a pre-built wheel installed via `pip` will suffice::

    pip install numpy==1.18.4

******************************************************************************
3. Confirm their bug exists
******************************************************************************

Their claim is simple enough: the wrong dtype is returned if one of the inputs
of the method `numpy.polymul` is a zero array. Open up a Python terminal and
type their code in and see if you can reproduce the bug::

    >>> import numpy as np
    >>> np.__version__
    '1.18.4'
    >>> a = np.array([1,2,3])
    >>> z = np.array([0,0,0])

>     >>> np.polymul(a.astype(np.int64), a.astype(np.int64)).dtype

Reviewer: Some information as to why the returned data types aren't correct might be helpful.

    dtype('int64')
    >>> np.polymul(a.astype(np.int64), z.astype(np.int64)).dtype
    dtype('float64')
    >>> np.polymul(a.astype(np.float32), z.astype(np.float32)).dtype
    dtype('float64')
    >>> np.polymul(a.astype(np.complex64), z.astype(np.complex64)).dtype
    dtype('complex128')

******************************************************************************
4. See if the bug still exists on the latest version of NumPy
******************************************************************************

Now that the bug is confirmed, you could try and solve it by opening a PR.
Although for this specific bug, it has already been solved 
`here <https://github.com/numpy/numpy/pull/17577>`_. To confirm,
first uninstall the old version of NumPy and install the latest one::

    pip uninstall numpy
    pip install numpy

Now run the code again and confirm that the bug was solved.

::

    >>> import numpy as np
    >>> np.__version__

>     '1.18.4'

Reviewer: This version number is the same as the one above where the bug exists. That can't be the case.

I would resolve this by noting in the running text of the doc that the version of numpy will be the latest, and that's what you should expect to see here.

.. _how-to-verify-bug:

##############################################################################
How to verify your first NumPy bug
##############################################################################

> This how-to will go through the process of verifying a bug_ that was submitted
> on the Github issues page.
> 
> .. _bug: https://github.com/numpy/numpy/issues/16354

Reviewer 2: ```suggestion
This how-to will go through the process of verifying a
`bug <https://github.com/numpy/numpy/issues/16354>`_ that was submitted
on the Github issues page.
```

    >>> a = np.array([1,2,3])
    >>> z = np.array([0,0,0])
    >>> np.polymul(a.astype(np.int64), a.astype(np.int64)).dtype
    dtype('int64')
    >>> np.polymul(a.astype(np.int64), z.astype(np.int64)).dtype
    dtype('float64')
    >>> np.polymul(a.astype(np.float32), z.astype(np.float32)).dtype
    dtype('float64')
    >>> np.polymul(a.astype(np.complex64), z.astype(np.complex64)).dtype
    dtype('complex128')
    Numpy/Python version information:
    >>> import sys, numpy; print(numpy.__version__, sys.version)
    1.18.4 3.7.5 (default, Nov  7 2019, 10:50:52) 
    [GCC 8.3.0]

******************************************************************************
1. Setup a virtual environment
******************************************************************************

Create a new directory, enter into it and set up a virtual environment using
your preferred method. For example, this is how to do it using `virtualenv`:

::

    virtualenv venv_np_bug
    source venv_np_bug/bin/activate

******************************************************************************
2. Install their version of NumPy
******************************************************************************

> If the bug was more complex, you may need to build the NumPy version they

Reviewer 3: I would suggest _"If the bug is more complex, you may need to.."_ or _"For complex bugs, you may need to..."_

referenced from source (insert link for how to build here). But for this bug,
a pre-built wheel installed via `pip` will suffice:

::

    pip install numpy==1.18.4

******************************************************************************
3. Confirm their bug exists
******************************************************************************

> So their claim is simple enough, they say that the wrong dtype is returned
> if one of the inputs of the method `np.polymul` is a zero array. Open up a
> python terminal and type their code in and see if you can reproduce the
> bug:

Reviewer 2: ```suggestion
Their claim is simple enough: the wrong dtype is returned 
if one of the inputs of the method `numpy.polymul` is a zero array. Open up a
Python terminal and type their code in and see if you can reproduce the
bug:
```

::

    >>> import numpy as np
    >>> np.__version__
    '1.18.4'
    >>> a = np.array([1,2,3])
    >>> z = np.array([0,0,0])
    >>> np.polymul(a.astype(np.int64), a.astype(np.int64)).dtype
    dtype('int64')
    >>> np.polymul(a.astype(np.int64), z.astype(np.int64)).dtype
    dtype('float64')
    >>> np.polymul(a.astype(np.float32), z.astype(np.float32)).dtype
    dtype('float64')
    >>> np.polymul(a.astype(np.complex64), z.astype(np.complex64)).dtype
    dtype('complex128')

******************************************************************************
4. See if the bug still exists on the latest version of NumPy
******************************************************************************

> Now that the bug is confirmed, you could try and solve it by opening a PR.
> Although for this specific bug, it has already been solved here_. To confirm,
> first uninstall the old version of NumPy and install the latest one:
> 
> .. _here: https://github.com/numpy/numpy/pull/17577
> 
> ::

Reviewer 2: ```suggestion
Now that the bug is confirmed, you could try and solve it by opening a PR.
Although for this specific bug, it has already been solved 
`here <https://github.com/numpy/numpy/pull/17577>`_. To confirm,
first uninstall the old version of NumPy and install the latest one::
```

(note that using two colons on the last line before the code block is enough and you don't have to add the double colon on a separate line)

    pip uninstall numpy
    pip install numpy

> 

Reviewer 2: Might be worth mentioning that if you're using venv or conda you can actually have parallel installations with different versions of NumPy.

Now run the code again and confirm that the bug was solved.

::

    >>> import numpy as np
    >>> np.__version__
    '1.18.4'
    >>> a = np.array([1,2,3])
    >>> z = np.array([0,0,0])
    >>> np.polymul(a.astype(np.int64), a.astype(np.int64)).dtype
    dtype('int64')
    >>> np.polymul(a.astype(np.int64), z.astype(np.int64)).dtype
    dtype('int64')
    >>> np.polymul(a.astype(np.float32), z.astype(np.float32)).dtype
    dtype('float32')
    >>> np.polymul(a.astype(np.complex64), z.astype(np.complex64)).dtype
    dtype('complex64')

******************************************************************************
5. What to do next
******************************************************************************

> Go to the NumPy issues page https://github.com/numpy/numpy/issues and see if
> you can confirm the existence of any other bugs. If you can, comment on the 
> issue saying so. Doing this helps the NumPy developers know if more than one
> user is experiencing the issue.

Reviewer 2: ```suggestion
Go to the `NumPy GitHub issues page <https://github.com/numpy/numpy/issues>`_
and see if you can confirm the existence of any other bugs. If you can, comment
on the issue saying so. Doing this helps the NumPy developers know if more than
one user is experiencing the issue.
```
